CRM/Signatures/Upgrader.php

```diff
@@ -0,0 +1,70 @@
+<?php
+/*------------------------------------------------------------+
+| SYSTOPIA Signatures                                         |
+| Copyright (C) 2017 SYSTOPIA                                 |
+| Author: B. Endres (endres@systopia.de)                      |
+|         J. Schuppe (schuppe@systopia.de)                    |
++-------------------------------------------------------------+
+| This program is released as free software under the         |
+| Affero GPL license. You can redistribute it and/or          |
+| modify it under the terms of this license which you         |
+| can read by viewing the included agpl.txt or online         |
+| at www.gnu.org/licenses/agpl.html. Removal of this          |
+| copyright header is strictly prohibited without             |
+| written permission from the original author(s).             |
++-------------------------------------------------------------*/
+
+use CRM_Signatures_ExtensionUtil as E;
+
+/**
+ * Collection of upgrade steps.
+ */
+class CRM_Signatures_Upgrader extends CRM_Signatures_Upgrader_Base {
+
+  /**
+   * Split the signatures settings into contact-specific settings.
+   *
+   * @return bool
+   *   Whether the upgrade task executed successfully.
+   */
+  public function upgrade_4200() {
+    $all_signatures = CRM_Core_BAO_Setting::getItem(
+      'de.systopia.signatures',
+      'signatures_signatures'
+    );
+    foreach ($all_signatures as $contact_id => $signatures) {
```

Comment:
You should take care of the case that $all_signatures is NULL, that will happen a lot

Reply:
Had to rewrite this part completely to not use `CRM_Core_BAO_Setting::getItem()` at all, as the old value may be corrupted and non-unserializable.

Instead, now retrieving the old setting via a manual query and trying to repair the serialization string to avoid data loss.

Also, signatures data is now being Base64 encoded before passing to `CRM_Core_BAO_Setting::setItem()` to avoid unserialization failures later on.